Show tag changes in object history view

iandees opened the pull request:
### Description

This adds highlighting of tag changes in each version to the history view for primitives:
- Tags that are **new** will be highlighted in green
- Tags that have **changed value** will be highlighted in yellow, with the value change described using an arrow like `old → new`
- Tags that have been **removed** will be highlighted in red, with the value `deleted`

Screenshot:
<img width="839" height="494" alt="image" src="https://github.com/user-attachments/assets/f6a066c5-84bb-4c84-b3fa-469238da6be8" />

This probably closes https://github.com/openstreetmap/openstreetmap-website/issues/738. I know there are other discussions and tickets related to this idea, but I couldn't find them in my few searches.

Ruby is not my native language, so please point out where I can improve the change.

### How has this been tested?

I ran the site locally in Docker, added some fake data to the database, and inspected it on my browser locally.

iandees commented:
> Unchanged tags look slightly odd with the white bar of a different width

Agreed. What do you think about adding a grey bar of the same width?

<img width="1065" height="612" alt="image" src="https://github.com/user-attachments/assets/f82bc84b-0d7f-4cd2-b90c-bfcbb2dd18aa" />

For the record, I suggested the change to the border color (rather than the whole cell or row) in an attempt to avoid having to deal with text on odd color backgrounds.

pablobm commented on app/views/browse/_common_details.html.erb:
```diff
@@ -38,4 +38,8 @@
 </div>
 
-<%= render :partial => "browse/tag_details", :object => common_details.tags %>
+<%= if defined?(@old_features) && @old_features
```
This shouldn't be necessary in Ruby. Instance variables (starting with `@`) resolve to `nil` if not defined.


```suggestion
<%= if @old_features
```

iandees replied:
Suggestion taken in 8b1fb0bf6.

pablobm commented on app/views/browse/_tag_details_with_changes.html.erb:
```diff
@@ -0,0 +1,18 @@
+  <h4><%= t ".tags" %></h4>
+  <div class='mb-3 border border-secondary-subtle rounded overflow-hidden'>
+    <table class='mb-0 browse-tag-list table align-middle'>
+      <% if defined?(@old_features) && @old_features %>
```
Same as above:

```suggestion
      <% if @old_features %>
```

pablobm reviewed (commented):
I love this ❤️ Some notes:
- The border in the middle (`border-start` of value cell) looks perhaps a bit too thin to me, particularly when contrasted with the one in the left (tag name cell). Probably not a big deal, but I wonder if other variants can be experimented.
- I see you added a background color for dark mode. To me it looks subtle enough not to clash with the text, so that's good. Still I wonder: is there a rationale for this difference?
- I wonder if there's a way to make this work with assistive technologies too. An `aria-*` tag or something, but I can't find anything right now and it's a very tricky subject anyway. Perhaps for a future PR.

tomhughes commented on app/helpers/browse_helper.rb:
```diff
@@ -72,6 +72,67 @@ def link_follow(object)
+      unless current_tags.key?(key)
+        changes[key] = { type: :deleted, previous: value }
+      end
+    end
```
Maybe `previous_tags.keys.difference(current_tags.keys).each` is better?

iandees commented:
> I see you added a background color for dark mode. To me it looks subtle enough not to clash with the text, so that's good. Still I wonder: is there a rationale for this difference?

I was experimenting but didn't mean to include this in the PR. I'll remove it.

> I wonder if there's a way to make this work with assistive technologies too. An aria-* tag or something, but I can't find anything right now and it's a very tricky subject anyway. Perhaps for a future PR.

I'm happy to explore this, but I agree that it should probably be another PR.

> The border in the middle (border-start of value cell) looks perhaps a bit too thin to me, particularly when contrasted with the one in the left (tag name cell). Probably not a big deal, but I wonder if other variants can be experimented.

Here's a variant where the whole row is highlighted in the color:

<img width="547" height="615" alt="image" src="https://github.com/user-attachments/assets/ad634f9b-f245-41e1-8d40-1ecb55406406" />

Highlight the row with a stronger left border:

<img width="421" height="759" alt="image" src="https://github.com/user-attachments/assets/32dbcac4-ad62-4f7c-b5c6-6e17d60f3a28" />

Here's one where just the value part of the table is highlighted:

<img width="436" height="786" alt="image" src="https://github.com/user-attachments/assets/326a2d91-8cb0-4908-85b1-e34e660505db" />

1ec5 commented:
This is a smart design, makes it a lot easier to find changes at a glance. My first impression is that we could probably even diff within a changed tag, word by word like MediaWiki does, for more clarity. But we can always fine-tune the presentation separately.

> I wonder if there's a way to make this work with assistive technologies too. An aria-* tag or something, but I can't find anything right now and it's a very tricky subject anyway. Perhaps for a future PR.

Along these lines, we should probably supplement the colors with some other visual distinction, because the traffic light colors aren’t great for colorblind friendliness. We already use these colors in the change counts in a changeset listing, but the user can depend on the position to distinguish between additions and deletions. Here the user can depend on deletions to be italicized and changes to have a “→”. But nothing distinguishes an added tag from an unchanged value. Not sure if that’s a problem with these particular colors. We should track colorblind friendliness in a separate issue if it isn’t trivial to deal with in this PR.

HolgerJeromin commented:
We could use the git/diff wording.
So add `+` and `-` In the first column to tackle color blindness. For changed values this could be empty but there we have the arrow.

We even could get rid of the arrow and be more like diff:
```
  amenity cafe
- name oldname
+ name newname
+ opening_hours sometimes
- fixme add opening time
```

pablobm commented on app/views/browse/_common_details.html.erb:
```diff
@@ -38,4 +38,4 @@
 </div>
 
-<%= render :partial => "browse/tag_details", :object => common_details.tags %>
+<%= render :partial => "browse/tag_details", :object => common_details.tags, :locals => { :tag_details_object => common_details } %>
```
How about renaming `tag_details_object` to `current_feature`? I had some difficulty reasoning about what it represented, due to the naming.

pablobm reviewed (commented):
Still looking into how to better resolve the duplication in `browse/_tag_details`, but these are some quick notes.

1ec5 commented:
<del>Strikethrough</del> is the default user agent styling for `<del>` tags. The corresponding `<ins>` tag gets <ins>underlined</ins>. Unfortunately, underlining gets confusing when text can be linked, and especially with `camel_case` keywords. We aren’t using **bold** for anything yet, and it would make it easier to see at a glance when something has been introduced in the element history.

pablobm commented on app/helpers/browse_helper.rb:
```diff
@@ -72,6 +72,63 @@ def link_follow(object)
+    previous_version = current_index&.positive? ? sorted_versions[current_index - 1] : nil
+
+    # Don't compare with redacted versions unless we're showing redactions
+    previous_tags = if previous_version && (!previous_version.redacted? || params[:show_redactions])
```
Would it make sense to filter out the redacted versions when finding the index above, and show changes bridging versions, pretending that the redacted ones don't exist?

So for example: if we have v1, v2(redacted), v3, then show changes between v1 and v3

pablobm commented:
Definitely, do as you please :+1: My branch was only an experiment.

iandees commented:
Ok, I think the code is probably in a decent spot now. I'm going to fiddle with trying to improve the visualization a bit now.

iandees commented:
> But why adding the +- to the value cell?

See conversation above. We are looking for something to show the change without relying solely on color.

iandees commented:
That looks better, but it's harder to distinguish between a tag's value changing and a tag getting added.

Maybe I'll change it so the key stays yellow but the value gets the red/green -/+ treatment.

iandees commented:
I switched to using an extra `<td>` for the +/- indicators and that got it to look the way I wanted:

<img width="417" height="692" alt="image" src="https://github.com/user-attachments/assets/003ebb91-7832-4479-8f4c-13cfd6321cc5" />

Happy to take feedback on this before spending much more time fixing tests.

iandees commented on app/helpers/browse_tag_changes_helper.rb:
```diff
@@ -0,0 +1,125 @@
+      [
+        tag.tr(:class => tag_change_class(change_info[:type])) do
+          safe_join([
+                      tag.th(format_key(key), :class => "diff-key-modified", :rowspan => 2),
```
Good question, I didn't know that was visible in the history view. I would propose removing it from the history view to reduce the complexity of this view. It would still be visible in the view of a single version.

iandees replied:
I made a change in fa4310f to remove the wikidata preview. It seems to work ... usually ... but it definitely makes the whole tag diff situation messier.

1ec5 commented:
I think @iandees has already responded to the outstanding visual feedback (see resolved review comments).

The only thing that might be debatable from a user standpoint would be the removal of Wikidata previews from history pages to keep the row-spanning key cell from getting mangled. If we think that’s a problem, it might be possible (if annoying) to split the key cell in half when showing the preview. Or the preview could occupy only the right-hand column on the history page. That said, the preview remains on version pages, so no information is actually lost, just slightly less convenient. We could ticket out restoring this functionality as tail work.

iandees commented:
Yep, the status of the PR is that the visual representation of tag history looks the same as the screenshot in [this comment](https://github.com/openstreetmap/openstreetmap-website/pull/6448#issuecomment-3443928967). The work that happened since is to fix lint/test/etc.

I'm happy to take more feedback, but it seems to be in a pretty good spot.

I think the wikidata lookup would probably be fairly easy to add back if that's important in this view.

pablobm commented on app/assets/stylesheets/common.scss:
```diff
@@ -1033,3 +1033,91 @@ img.trace_image {
+  }
+
+  .tag-modified {
+    background-color: rgba(var(--bs-warning-rgb), 0.08) !important;
```
Are all these `important!` declarations necessary? I just tried removing all of them, and the only change I could notice was taller column rows:

Current site | This PR | Removed `important!`
--- | --- | ---
<img width="406" height="281" alt="Tags in changeset, as currently on production" src="https://github.com/user-attachments/assets/4b3e94c3-b174-4e5c-b410-5ecdedd0cbaf" /> | <img width="406" height="281" alt="Tags in changeset as in this PR, with differences highlighted" src="https://github.com/user-attachments/assets/935a98cd-2663-4e92-a34c-0b4277a5d5e6" /> | <img width="406" height="281" alt="Tags in changeset, as in this PR, but columns are a bit taller" src="https://github.com/user-attachments/assets/e5c33dfd-695b-48e0-9ebf-20a3f12b7dfb" />

Mentioning because `important!` is frowned upon in CSS circles.

pablobm reviewed (commented):
I'm seeing two things when comparing the production and the PR renders side by side. This is for the normal list of current tags, as opposed to the history view with diffs:

Production | PR
--- | ---
<img width="411" height="401" alt="Tags show with a grey background behind the keys, while values have a white background. Wikidata row has a 'barcode' icon to show further info" src="https://github.com/user-attachments/assets/c27eab60-e593-4fef-a1fc-c0001e16ece2" /> | <img width="411" height="401" alt="Tags show with an all-white background. There's no 'barcode' icon in Wikidata rows" src="https://github.com/user-attachments/assets/0e092f9f-5167-4b46-8c2b-30f2e87f0b7c" />

On production, keys have a grey background and values a white background. On this PR it's an all-white background. (I'm not too bothered by this one, but I don't know what others think and whether they are aware).

On production, `wikidata` tags show the "barcode" icon to trigger an expand for more info. This is missing on the PR version.

iandees commented:
I've removed the `!important` flags and replaced them with some more specific CSS. This also means the other tag views aren't affected by the change to the tag history view.

I also made a small change to the Wikidata Javascript so that it works with the tag diff, so I added that back.

The row heights are fixed as well. This one was a bit confusing to me because it looked great locally, but I discovered I have to rebuild CSS and restart the docker compose system for every change to the CSS to see all changes.

This is where we're at:

<img width="371" height="488" alt="image" src="https://github.com/user-attachments/assets/c4fab22e-ae46-478d-a806-59abcc5abecf" />

1ec5 commented:
Yeah, the only way to make the layout work would be to temporarily (?) split and duplicate the key cell so that it no longer spans two rows, then insert the preview row immediately beneath the row that the user clicked on. But that would break the visual link between the old and new values for that key. I don’t think getting this feature to work on the history page should be a blocker, since the preview is only a convenience and one that hasn’t been around for very long. The user can still access all the same information on Wikidata by clicking on the QID. We can ticket out restoring the feature to this page.

iandees commented:
Alrite, I removed the Wikidata preview in the history/tag diff view while keeping it in the other tag views.

Also, I believe I found the missing class that was causing the table cells to get a couple extra pixels of top and bottom padding.

waldyrious commented:
> ~Strikethrough~ is the default user agent styling for `<del>` tags. The corresponding `<ins>` tag gets underlined. Unfortunately, underlining gets confusing when text can be linked, and especially with `camel_case` keywords. We aren’t using **bold** for anything yet, and it would make it easier to see at a glance when something has been introduced in the element history.

Apologies for the late comment but I don't see additional discussion about this. I am confused about the `<del>`/`<ins>` approach having been abandoned, since the removed parts are pretty evident with the strikethrough formatting, and as for the new parts, CSS could be used to omit the underlining, since the absence of strikethrough should be enough to distinguish the old vs. new values.

I'm especially puzzled that this additional marking of the values (besides the coloring), which is being considered due to accessibility concerns, has moved away from the semantic HTML tags that would make the result universally unambiguous not just for those with color vision limitations, but also for the blind users relying on screen readers. (And it would even be machine-readable!)

1ec5 commented:
> the unchanged values appear with a grey background

This means a version that only changes geometry will have a completely grayed-out table. I think it could help visually differentiate the history page from the version page.

> Apologies for the late comment but I don't see additional discussion about this. I am confused about the `<del>`/`<ins>` approach having been abandoned, since the removed parts are pretty evident with the strikethrough formatting, and as for the new parts, CSS could be used to omit the underlining, since the absence of strikethrough should be enough to distinguish the old vs. new values.

Do you suggest using `<del>` and `<ins>` tags in addition to the + and − indicators? Or do you mean we should return to inlining a changed tag, using `<del>` and `<ins>` to distinguish old and new values? Inlining could be useful if we later introduce per-word or per-character diffing, to make subtle changes easier to spot. If so, I agree with using CSS to nix the underline.

Without that more granular diffing, this more structured design would be more effective at highlighting subtle changes. If we keep it, we should still make the + and − indicators more screen reader accessible: wrap each of these symbols in a `<abbr title>` tag that the screen reader can expand to something like “Removed” or “Added”. (The tooltip text should be localizable.)

waldyrious commented:
> This option wasn't abandoned or moved away from. It was never attempted.

Ah, my bad — I totally overlooked the author of [this comment](https://github.com/openstreetmap/openstreetmap-website/pull/6448#issuecomment-3410364736) :sweat_smile:

(And responding to your question, my suggestion would be precisely that one — not sure if implementing it would bring more value than the mock-up in that comment, though, at least until a decision is made about whether it's desirable, or concrete use cases where its usage may not be optimal are identified.)

> Do you suggest using `<del>` and `<ins>` tags in addition to the + and − indicators? Or do you mean we should return to inlining a changed tag, using `<del>` and `<ins>` to distinguish old and new values? 

IMHO the latter should be sufficient, since for regular vision users the colors alone would convey the necessary information, and that way we'd avoid the extra visual complexity.

> Inlining could be useful if we later introduce per-word or per-character diffing, to make subtle changes easier to spot.

Agreed, but to be clear I'd argue it'd still be useful with line-based diffing.

> Without that more granular diffing, this more structured design would be more effective at highlighting subtle changes.

By "more structured design" do you mean the +/- indicators? And if so, can you explain what is it that makes it more effective at highlighting subtle changes? I'm not sure I follow.

1ec5 commented:
> By "more structured design" do you mean the +/- indicators? And if so, can you explain what is it that makes it more effective at highlighting subtle changes? I'm not sure I follow.

Because the value cells are left-aligned (or right-aligned in a right-to-left environment), showing the new value under the old value can help the user to spot any minor changes just from the characters being misaligned between the two cells. For example, [this relation](https://www.openstreetmap.org/relation/18322858/history/10)’s fairly long `description=*` is edited to replace “shift” with a naughty word one letter shorter, then showing the full new value under the full old value would make that change more apparent. But showing both the old and new values together on one line in their entirety, separated by a →, would make it more difficult to spot that one-letter change. In the future, an inline diff word by word would also make the change apparent, so we could go back to combining the old and new values in a single cell.

1ec5 commented:
Strikethrough can somewhat obscure a <del>- hyphen</del>, <del>– dash</del>, or <del>− minus</del> just like a Chinese <del>一 one</del> or Japanese <del>ー long vowel mark</del>. It might be more of a problem on 1× displays than 2× displays, where the strikethrough can be half as thick. I’m less concerned about it than underlines, since an <ins>_ underscore</ins> has a special role in OSM tagging and we tend to [`link_values`](https://github.com/openstreetmap/openstreetmap-website/pull/6448#issuecomment-3556958546) with underscores in them.

I guess there’s a potential caveat if someone tries to print out the page, since we’re relying on the cell backgrounds to provide contrast against the ± indicators. Strikethrough could help with that and the case where someone disables the stylesheet. But I’m pretty sure we haven’t given these use cases much consideration lately. If the only concern is about screen reader behavior, then we could wrap the old value in `<del>` and the new value in `<ins>` and use CSS to eliminate any styling of both tags in this table. Otherwise, I would view the suggestion for strikethrough as an alternative to the ± indicators, because if we do both then we’re starting to really overengineer the table with four visual differences between old and new.

pablobm commented:
For completeness, I just realised that we use strikethrough when showing the features removed in a changeset:

<img width="405" height="332" alt="List of ways and nodes affected in a changeset, where a way is rendered striken-through (is that correct English?)" src="https://github.com/user-attachments/assets/8361fe1d-bdca-4688-8fd4-9118f8013c18" />

Differences:
- From my experiments, it appears to only show the feature id number and version. This avoid the issue with Chinese/etc scripts.
- Strikethrough is done with an `<s>` tag rather than a `<del>`. I don't see a counterpart `<ins>` or equivalent in added features.

waldyrious commented:
> If the only concern is about screen reader behavior, then we could wrap the old value in `<del>` and the new value in `<ins>` and use CSS to eliminate any styling of both tags in this table.

I do agree that using the semantic tags with the usual CSS striped out would be desirable, since their meaning would be conveyed directly in screen readers, and by the colors (as progressive enhancement) for those with normal vision.

> Otherwise, I would view the suggestion for strikethrough as an alternative to the ± indicators, because if we do both then we’re starting to really overengineer the table with four visual differences between old and new.

I'd be inclined to share that view, but I'd like someone who's actually affected by color issues to comment on whether this is an acceptable solution. I mean, if we do hide the visual signals of `<ins>` and `<del>` in normal rendering, then we'd be leaving them with nothing but color to distinguish the values. We could assume that people who are affected by red/green styling would be likely to have custom user stylesheets or browser extensions to enhance color-only encoding of diffs, since that's a pretty common pattern; but I wouldn't want to take that decision (to deem color-only distinction sufficient in non-screen-reader contexts) on their behalf.

By the way, you raised a good point about e.g. printing the page — I'd suggest that matching rules be set up in [`print.scss`](https://github.com/openstreetmap/openstreetmap-website/blob/master/app/assets/stylesheets/print.scss) to either restore the strikethrough styling (and possibly also underline, though maybe using an alternative CSS property like `border-bottom` to shift it downwards from the text baseline), or use pseudo-elements to prepend `+` and `-` indicators, when color is absent.

Anyway, we're pretty much veering towards bikeshedding now, so just to be clear, I'd be pretty happy with something like [this](https://github.com/openstreetmap/openstreetmap-website/pull/6448#issuecomment-3489346156) (with grey for unchganged values, as shown [here](https://github.com/openstreetmap/openstreetmap-website/pull/6448#issuecomment-3516849705)), as long as the changed values are wrapped in `<ins>`/`<del>` (regardless of their visibility).

mmd-osm commented:
> It remains possible to delete an object without deleting the tags using the raw API

That’s not correct, the API always deletes keys in this case. The client has no way to influence this behavior. I don’t remember what the old AMF code for Potlatch was doing, but since it’s gone for a long time already that shouldn’t really matter anymore.

mmd-osm commented:
You can add these tags in osmChange, which is more of a curtesy to human readers to make clear what is being deleted. The API will ignore it, though. —> https://wiki.openstreetmap.org/wiki/OsmChange

iandees commented:
> Could we try and clean up the commit history here

Yep, absolutely.

iandees committed:
Refactor BrowseTagsHelper to extract TagParser

iandees committed:
Add skip_wikidata_preview option to BrowseTagsHelper

iandees committed:
Add BrowseTagChangesHelper for calculating tag diffs

iandees committed:
Show tag changes in object history view

iandees committed:
Style the tag history view

pablobm reviewed (commented):
I think this is ok. The initial refactor and the markup-generating code make things difficult to follow, but it can be followed a bit better when going commit-by-commit.

If anything, the markup generation could be untangled a bit. I gave it a shot at https://github.com/iandees/openstreetmap-website/compare/highlight-tag-changes-in-history...pablobm:openstreetmap-website:highlight-tag-changes-in-history and I think it's an improvement there. Not sure if it worth pursuing elsewhere, after playing with it a bit more.

iandees committed:
Clean up browse tags helper code

tomhughes commented on app/helpers/browse_tag_changes_helper.rb:
```diff
@@ -0,0 +1,134 @@
+      tag.del(format_value(key, change_info[:previous] || "", :skip_wikidata_preview => true))
+    when nil
+      # For unknown versioning show the current value with the wikidata preview
+      format_value(key, change_info[:current] || "", :skip_wikidata_preview => false)
```
Same here for the current value? As far as I can see it should always have one?

tomhughes commented on app/helpers/browse_tag_changes_helper.rb:
```diff
@@ -0,0 +1,134 @@
+      # Only add diff indicator cell if we're in a history/diff context
+      cells << tag.td(get_change_indicator_text(change_info[:type]), :class => get_indicator_cell_class(change_info[:type])) if change_info[:type]
+
+      cells << tag.td(format_tag_value_with_change(key, change_info), :class => "py-1")
```
Why break this down into multiple statements that append to a temporary array when the modified version just does it all inline?

tomhughes reviewed (commented):
In addition to the specific comments could you rebase this please as there are apparently some conflicts.

Having now tried this after reviewing the code I think a lot of my comments may be explained by the fact that this only works in the history list, so I guess `current_version` is not the current version in the normal sense but is the version currently being rendered?

I have to say I find it a bit odd that the changes aren't shown when looking at a specific version - it actually took me a while to find out how to trigger this to the extent that I was putting in debugging code to try and find out why it didn't seem to be working!

1ec5 commented:
> I think moving the breadcrumb and pagination bar to the top of the page would address this confusion.

#6759

> I also wonder if we should have a separate page for diffing two specific versions.

#6761

iandees committed:
Address review feedback on BrowseTagChangesHelper

iandees commented on test/helpers/browse_tags_helper_test.rb:
```diff
@@ -90,6 +90,7 @@ def test_format_value
     html = format_value("website", "example.com/page")
     assert_dom_equal "example.com/page", html
   end
+<<<<<<< HEAD
```
Damnit, yes.

tomhughes reviewed (commented):
A couple of overall comments...

Firstly, please don't create "address review feedback" commits but instead fold the changes into the original commit(s) where the code being changed was introduced.

Secondly, maybe it's just me but to me the way this shows changes seems to be backwards to me... Consider this example:

<img width="421" height="1182" alt="Image" src="https://github.com/user-attachments/assets/c04f5b3f-ed49-4a08-ab44-56c2b4ea784d" />

That way was `leisure=park` in v1 and changed `amenity=school` to in v2 but it looks like the change is happening in v1, which is logically impossible and hence why I spotted it. Then you realise it's actually showing the change backwards as well at which point it makes a bit more sense but is assuming you're reading backwards in time from the present version.

Maybe it's just the programmer in me and normal people expect it to work like that but it's not what a programmer used to reading diffs expects to see!

stillhart commented:
I wanted to mention that I tried to solve the same but went with a different approach, see #6995. If mine doesn't make it, it might be of interested anyway.

stillhart commented:
Honestly, I prefer my approach a lot more. It's way simpler, clear, doesn't break and especially doesn't confuse with existing. The diff shouldn't be in the state. The diff is the diff in between states. It's whats changed, not what is. Having the diff in the state at a given time makes it more ambiguous whether we look forward or backward. It's something to look into, not something that must be visible at all times. 

Further, having just both values above/below each other makes it clearly comparable by eye. 

Lastly, the KISS bootstrap colors fit all existing schemas. 

It looks like this in the existing view
<img width="347" height="1134" alt="image" src="https://github.com/user-attachments/assets/4e434e8b-26e1-451c-bccc-82a2eff8f7df" />

and once expanded using no js it's all in there. 

<img width="374" height="359" alt="image" src="https://github.com/user-attachments/assets/100e035c-d061-4969-903c-0f8fa07ae730" />
